polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -641,6 +713,247 @@ MemoryAccess<OpTy>::classifyMemoryAccess(DataFlowSolver &solver) const {
+}
+
+template <typename T>
+void MemoryAccessAnalysis::build(T memoryOp, DataFlowSolver &solver) {
+  static_assert(llvm::is_one_of<T, AffineLoadOp, AffineStoreOp>::value);
+
+  MemRefType memRefType = memoryOp.getMemRefType();
+  if (!memRefType.getLayout().isIdentity())
+    return;
+
+  // Currently we are only interested in SYCL memory accesses that have:
+  //  - base operand equal to the result of a sycl accessor subscript, and
+  //  - a single index with zero value.
+  MemRefAccess access(memoryOp);
+  if (!isa<sycl::SYCLAccessorSubscriptOp>(access.memref.getDefiningOp()) ||
+      !hasZeroIndex(access))
+    return;
+
+  LLVM_DEBUG(llvm::errs() << "Candidate op:" << memoryOp << "\n");
+
+  auto accessorSubscriptOp =
+      cast<sycl::SYCLAccessorSubscriptOp>(access.memref.getDefiningOp());
+  Value accSubIndex = accessorSubscriptOp.getIndex();
+
+  // Try to determine the underlying value(s) of the accessor subscript index
+  // operand. The number of underlying values should be equal to the
+  // dimensionality of the sycl id used as an index in the accessor subscript
+  // operation. Example:
+  //   sycl.constructor @id(%id, %i, %j) : (memref<?x!sycl_id_2>, i64, i64)
+  //   %sub = sycl.accessor.subscript %acc[%id] ...
+  // Here the underlying values for '%id' are {%i, %j}. The number of underlying
+  // values matches the dimensionality of the sycl id.
+  SmallVector<Value> underlyingVals = getUnderlyingValues(
+      accessorSubscriptOp.getOffsetOperandIndex(), accessorSubscriptOp, solver);
+  if (underlyingVals.empty()) {
+    LLVM_DEBUG(llvm::dbgs() << "Failed to resolve underlying value(s) for: "
+                            << accessorSubscriptOp.getIndex() << "\n");
+    return;
+  }
+
+  assert(
+      sycl::getDimensions(accSubIndex.getType()) == underlyingVals.size() &&
+      "Number of underlying values should be equal to dimensionality of the id "
+      "used to index the accessor");
+  LLVM_DEBUG({
+    for (Value val : underlyingVals)
+      llvm::dbgs() << "Underlying val: " << val << "\n";
+  });
+
+  // Construct the memory access matrix. The number of rows is equal to the
+  // dimensionality of the sycl.id used by the accessor subscript operation. The
+  // number of columns is equal to the number of loops surrounding the memory
+  // access plus (TODO) the dimensionality of the sycl item received by the
+  // kernel.
+  SetVector<Operation *> enclosingLoops =
+      getParentsOfType<AffineForOp>(memoryOp->getBlock());
+  MemoryAccessMatrix accessMatrix(sycl::getDimensions(accSubIndex.getType()),
+                                  enclosingLoops.size());
+
+  // Collect the loop induction variables.
+  // TODO: collect also the global thread ids.
+  SmallVector<Value, 4> IVs;
+  for (Operation *loop : llvm::reverse(enclosingLoops))
+    IVs.push_back(cast<AffineForOp>(loop).getInductionVar());
+
+  OpBuilder b(access.memref.getContext());
+  Value zero = b.create<arith::ConstantIndexOp>(b.getUnknownLoc(), 0);
```

Comment:
Are you aware of when will this value be freed? I assume it will outlive the matrix as it's owned by the context?
If that's the case, that's alright, although we might want to make it share the same lifespan as the matrix itself in the future (`OwningOpRef`). It's fine for this PR, though.

Reply:
It will outlive the matrix. I'll think about using OwningOpRef.